crates/bevy_app/src/app.rs

```diff
@@ -57,14 +58,42 @@ pub struct App {
     /// A container of [`Stage`]s set to be run in a linear order.
     pub schedule: Schedule,
     sub_apps: HashMap<AppLabelId, SubApp>,
+    /// Whether or not the App has been run and/or updated yet.
+    #[cfg(debug_assertions)]
+    pub has_run: bool,
 }
 
 /// Each `SubApp` has its own [`Schedule`] and [`World`], enabling a separation of concerns.
 struct SubApp {
     app: App,
     runner: Box<dyn Fn(&mut World, &mut App)>,
 }
-
+#[cfg(debug_assertions)]
+impl Drop for App {
+    /// Gives a warning if the [`App`] is [`Drop`]ped without being run.
+    fn drop(&mut self) {
+        // Checks if the app has run yet
+        if !self.has_run {
+            let message = "You have not called `App::run` or `App::update`. Any systems and/or plugins added to this `App` will not run.";
+            // `get_default` lets you access the current `Subscriber`.
+            // Bevy's `LogPlugin` will initialize a `Subscriber`, so this lets us check if logging is being used.
+            dispatcher::get_default(|d| {
+                #[cfg(feature = "tracing")]
+                {
+                    let _ = info_span!("bevy_app::App::drop").enter();
+                }
```

Comment:
as you create a scope here, the span is immediately dropped.
I'm not sure the span is useful anyway as there would only be a log message in it

Reply:
Thanks for pointing this out!